### PR TITLE
ENH: Fix Polynomial.fit() for multiple polynomial fitting

### PR DESCRIPTION
**Pull request by wrzadkow**
Fixes multiple polynomial fitting bug (Issue #4554 and Issue #16380). If `y.shape` is `(M, K)`, a list of `Polynomial` objects is returned, as advertised. Adds tests and updates the docstring accordingly.

### CODE REVIEW

File: numpy/polynomial/tests/test_classes.py [hunk 1]
```diff
@@ -145,14 +145,23 @@ def test_fit(Poly):
...
-    y = f(x)
+    y, z = f(x), g(x)
+    yz = np.column_stack((y,z))
 
     # check default value of domain and window
     p = Poly.fit(x, y, 3)
     assert_almost_equal(p.domain, [0, 3])
     assert_almost_equal(p(x), y)
     assert_equal(p.degree(), 3)
+    [q,r] = Poly.fit(x, yz, 3)
```
Comment (rossbar) [inline numpy/polynomial/tests/test_classes.py:159]:
```suggestion
    q, r = Poly.fit(x, yz, 3)
```
I think a tuple for the outputs is fine here.

Also, a bit of a nit on the variable naming - I'm not sure I like `q` and `r` given their "special meaning" in linear algebra (i.e. QR factorization). Maybe `p_y` and `p_z` instead? Not a blocker, just a thought.

**Review by rossbar — CHANGES_REQUESTED** [review]
Thanks @wrzadkow! If the discussion in #16380/here converges to a consensus that a list of series instances is the preferred output, the work here seems like a nice implementation of that option. 

I haven't yet gone through it in detail (particularly the tests), but did notice a few things. Hopefully this PR can help spur the discussion on what the preferred output would be for this use-case.

**Comment by wrzadkow** [comment]
Many thanks, @rossbar ! I support `fit()` returning instances of `Polynomial` class, hence such implementation in this PR, but of course I'm ready to adapt this solution if the discussion converges elsewhere. 

In the meantime, I implemented all your suggestions, thanks a lot for them.

**Comment by rossbar** [comment]
> I support fit() returning instances of Polynomial class, hence such implementation in this PR, but of course I'm ready to adapt the code if the discussion converges elsewhere.

I tend to agree - it would seem inconsistent for `Polynomial.fit` to return an instance in the `K=1` case, but a coefficient array in the `K > 1`. Plus the `*fit` functions from the corresponding modules (e.g. `polynomial.polyfit`) can be used to get coefficient *arrays* rather than class instances.

Either way, the current documentation is definitely out-of-sync with the implementation, so some sort of decision will have to be arrived at so we can fix this issue.

**Comment by rossbar** [comment]
I'm going to close this now that #16506 is merged - thanks @wrzadkow for taking the time to put this together! I think it was helpful in framing the discussion by having a concrete example of one of the options. If the consensus in the related discussion shifts this can be revisited.

**Pull request closed without merging**